### ISSUE TITLE
MutationObserver needs an equivalent of disconnected & connected callbacks

### ISSUE THREAD

**Issue by rniwa** [whatwg/dom#533]
It would be useful to have a way of observing when a node gets connected or disconnected from a document similar to the way `connectedCallback` and `disconnectedCallback` work for custom elements but unlike those at the end of the microtask like all other mutation records.

**Comment by rniwa** [comment]
@smaug---- @justinfagnani

**Comment by domenic** [comment]
It wasn't clear from the original thread what the use cases were. @pemrouze just said it would be nice, then people started discussing how. We never even talked about why they couldn't use custom elements.

I agree that if we had a need for this ability extending MutationObserver would be the way forward but I'd like some more justification in the form of concrete use cases that are hard to accomplish with custom elements.

**Comment by rniwa** [comment]
Agreed. It's always good to have concrete use cases before adding new API.

**Comment by smaug----** [comment]
IIRC this was discussed around the time when MutationObserver was designed and the idea was that since one can add childList observer to the document and then just check document.contains(node), that should in principle be enough (and the things to observe was kept in minimum).
However Shadow DOM wasn't really a thing at that time, and that may have changed the needs.
Concrete use cases, which can't be solved with Custom Elements, would be indeed nice.
The more we add APIs, the more complicated platform we have to maintain and adding more and more API tends to slow down implementations, or at least makes it harder to optimize them.

**Comment by justinfagnani** [comment]
This would be useful for polyfilling custom elements, but obviously that's not a good use case for why custom elements aren't enough.

If I can page in my requests around MutationObservers one had to do with observing access shadow boundaries so we don't have to patch attachShadow, which might be relevant here, and have some way of unobserving nodes outside of disconnect(), say when the nodes are disconnected.

**Comment by pemrouz** [comment]
> It wasn't clear from the original thread what the use cases were. @pemrouze just said it would be nice, then people started discussing how. We never even talked about why they couldn't use custom elements.

* A framework that does some housekeeping when a component is removed (e.g. tearing down subscriptions). The framework is outside the component, so would have to rely on every component emitting this event manually which isn't feasible/reliable. The `disconnectedCallback` is the current solution, but it requires wrapping every component in a wrapper component. This has other benefits (e.g. dynamic registry), but the less that can be done at the framework level would be great and this seemed generic enough to enquire about.

* It would be nice for a component to be able to declaratively manage subscriptions it's own subscriptions, rather than have to emit the event manually or be forced to imperatively manage the subscription (see original thread).

**Comment by smaug----** [comment]
Wouldn't a childList observer in document take care of those use cases, at least in common cases?

**Comment by pemrouz** [comment]
It's unfortunately [notoriously difficult](https://stackoverflow.com/questions/44935865/detect-when-a-node-is-deleted-or-removed-from-the-dom-because-a-parent-was) to do with MutationObserver, and pretty much guaranteed to be buggy (due to it being batched) and inefficient (having to watch every change on the entire document).

**Comment by WebReflection** [comment]
> This would be useful for polyfilling custom elements

polyfills for custom elements already work and are either based on MutationObserver or DOMNodeInserted/DOMNodeRemoved with a lookup for registered elements in the list of changed nodes.

Using polyfills is a weak reason, specially because anythig new needs to be polyfilled too.

**However** I do have a use case in hyperHTML too but I believe React and many other frameworks would love this feature as well.

CustomElements mechanism to understand their life-cycle are excellent but not everyone writes Custom Elements based sites and knowing a generic form, input, list element, grid card, has been added/removed would be ace.

in hyperHTML I do have control of each vanilla DOM component but I need to do this "dance" per each event:
```js
function dispatchTarget(node, isConnected, type, e) {
    if (components.has(node)) {
      node.dispatchEvent(e || (e = new Event(type)));
    }
    // this is where it gets a bit ugly
    else {
      for (var
        nodes = node.children,
        i = 0, length = nodes.length;
        i < length; i++
      ) {
        e = dispatchTarget(nodes[i], isConnected, type, e);
      }
    }
    return e;
  }
```

The dispatching after recursive search is done per each `record.addedNodes` and `record.removedNodes` but if I had a platform way to loop over all known connected/disconnected nodes my code, and my life, would definitively be easier.

**Comment by pemrouz** [comment]
@domenic any more thoughts?

**Comment by rniwa** [comment]
Here's a concrete use case. A editor library wants to know when the editor lost the focus. Because `blur` event does not fire on an element that got disconnected from a document before losing the focus, we'd need to monitor when the element is disconnected. To do this, we currently need to monitor all child node changes of the parent node, and then detect when the editor node is removed. With the proposed observation type, we can observe the disconnection of the editor element instead.

**Comment by WebReflection** [comment]
@rniwa what you described looks more like a `blur` related bug (either browser or specs) but I'm happy you found a concrete use case.

Another one I have these days is a declarative way for an element inside a template to perform setup operations without needing its container to know about these

```html
<tempalte>
  <div>
    One <span>or more</span> elements and then
   <p
    class="timer"
    onconnected="this.textContent = (new Date).toLocaleString()"
  > time </p>
  </div>
</template>
```

**Comment by smaug----** [comment]
That example doesn't use MutationObserver at all. onconnected sounds like a Mutation Event.

**Comment by rniwa** [comment]
`blur` event is not supposed to fire when an element is removed. Due to security concerns, we can't fire `blur` synchronously in that case.

**Comment by WebReflection** [comment]
@smaug you're right, that example is not directly related but it's something I can do already and it's based on MutationObserver. The `onconnected` is just a hint, imagining a list of added nodes per each record could be checked against `node.hasAttribute("onconnected")` and react accordingly.

That being said I wouldn't mind having a native `onconnected/disconnected` event in the DOM though but that's another story.

- - -

@rniwa 

> Due to security concerns, we can't fire blur synchronously in that case.

I am having hard time understanding why there wouldn't be security concern if `blur` can be intercepted through an observer ... what's the difference?

A focused node that never fires blur is a source of troubles. Mitigating troubles through other mechanism is a work around, not a solution to that issue.

Regardless, if that's the case you were looking for to justify connected/disconnected, I'm OK with it.

**Comment by rniwa** [comment]
`MutationObserver` fires at the end of the current micro-task, that's much later than synchronous timing which poses a security threat in our engine.

**Comment by wessberg** [comment]
Yes, this is much needed indeed.
Tracking when elements are connected/disconnected is rather easy with MutationObservers when *not* using Shadow DOM, but with it and as I see it, we have to observe the childList of the root of all parents going recursively up the tree from the node in question.

**Comment by caridy** [comment]
I would like to revamp this conversation. It seems that we have more use-cases that can help us to come to an agreement about this feature request. Maybe adding it to the agenda for the F2F meeting.

**Comment by rniwa** [comment]
Yeah, I think we should just add this. Here's a concrete proposal:

Add boolean `connectedness` to [`MutationObserverInit`](https://dom.spec.whatwg.org/#dictdef-mutationobserverinit), which if set would make the mutation observer start receiving a mutation record whenever the target node is connected to or disconnected from a document.

Add two booleans `becomeConnected` and `becomeDisconnected` to `MutationRecord`. `becomeConnected` is set to true when the target node is newly connected to a document. It's set to false otherwise. Similarly, `becomeDisconnected` is set to true if the target node is newly removed from a document. It's set to false otherwise.

`subtree` option in mutation observer would the observation subtree-wide.

I think there is an option to coalesce the record with the one which inserts / removes the node but a simpler model is to simply create a new mutation record for each.

**Comment by wessberg** [comment]
Sounds great. Wouldn't it be sufficient with a simple boolean `connected` which is `true|false`?

**Comment by rniwa** [comment]
> Sounds great. Wouldn't it be sufficient with a simple boolean `connected` which is `true|false`?

Sorry, I wasn't thinking through. We could just add new mutation record types like `connected` / `disconnected` and not even add a boolean.

**Comment by trusktr** [comment]
> We never even talked about why they couldn't use custom elements.

@domenic The main use case I imagine is with builtins, so we can get connected/disconnected behavior that behaves the same as the callbacks for Custom Elements do.

It'd be useful in some of the other issues I've had troubles with.

> It's always good to have a list of concrete use cases before adding new API.

@rniwa Another use case is simply: someone else made a web app, now we want to manipulate it without necessarily touching their source code (maybe we don't have access to the source, maybe we are writing a browser plugin, etc).

In a sense the purpose of this feature is more similar to the purpose of jQuery: manipulate existing DOM elements with it. This is in contrast to the act of designing the behaviors of our own DOM elements (custom elements), instead we can use this to manipulate existing DOM elements (builtin or custom).

👍 to this.

**Comment by caridy** [comment]
@rniwa here is our update on this:

today, there is no way to observe the synchronous connection and disconnection of a particular element, the only option is MutationObserver (which is async), or registering a custom element (which sometimes it is not an option or overkill).

Just to clarify, in the meeting @domenic mentioned Mutation Events, (e.g.: `DOMNodeInserted` works in all browsers today but does NOT work inside a shadow, and doesn't have the same semantics as connectedCallback). Additionally, `DOMNodeInsertedIntoDoc` does have the same semantics, but does not work in all browsers, neither inside a shadow).

The only real solution today is a very hacky solution, same solution used by us and by the polymer custom element polyfill, which is to patch all DOM APIs that allow you to mutation the childNodes, so you can insertion and removal of elements, to simulate the synchronous connect/disconnect callback semantics.

My question is, can we get a reliable way to listen for this node reaction, synchronous, without registering the element?

**Comment by wessberg** [comment]
@caridy, I'm not sure if I believe a synchronous approach will be able to move forward since we had that with Mutation Events (which as you say doesn't work reliably with Shadow roots) and one of the primary reasons to move away from them in favor of MutationObserver was their synchronous nature. I'm certain you have perfectly valid reasons to want a synchronous solution, and I agree that this is an important missing piece in the web platform right now, but if we can find a way to move forward with an asynchronous approach and with an API that is similar to other related observers such as MutationObserver, I think the odds of getting implementors behind this is greater. I've experimented with this and created a library called [ConnectionObserver](https://github.com/wessberg/connection-observer) which batches connection entries together in microtasks, and I'd like to propose something similar to that.

**Comment by annevk** [comment]
@wessberg how is https://github.com/whatwg/dom/issues/533#issuecomment-405530738 solved by extensions to `MutationObserver`? You'd still not know about shadow tree mutations.

Having reread the thread it's not entirely clear to me everyone is on the same page here when it comes to use cases and requirements, which might also be why we keep hearing different solutions being proposed.

Starting from the top at https://whatwg.org/faq#adding-new-features would be beneficial I think.

**Comment by wessberg** [comment]
@annevk, agreed. That is why I propose an entirely new observer as described in my previous comment.

**Comment by caridy** [comment]
@annevk I'm not proposing any particular solution, just trying to describe why we need this, and what are the options today that do not fulfill those use-cases.

To be more specific about use-cases, I have one main user-story for the sync version of the connect-ness detection:

> as a developer, I can emulate what you can do with a custom element callbacks without requiring to use a custom element.

The use-cases are various, here are a few:

* frameworks and libraries that do not use custom elements, requires a recursive search on subtrees to trigger their own disconnect hook even though not all elements use that hook.
* frameworks and libraries that create node-trees during the diffing algo, and later on install them at once into the document, will have to either a) trigger the connect hook when the element is added to its parent (even though it is not really in the doc yet - usually this is the solution), or b) recursively search in the virtual tree for vnodes that requires connect hook, to trigger that after the root was inserted into the dom, or c) never do node-trees in memory, always create them in the actual document (which has perf implications).
* polyfilling custom elements, or emulating custom elements in general, is impossible without hacking many APIs today.

**Comment by annevk** [comment]
So by synchronous you mean equivalent timing to custom elements, not equivalent timing to mutation events? Why is mutation observer timing insufficient?

Why would you need to polyfill/emulate custom elements at this point?

**Comment by caridy** [comment]
@annevk 

> So by synchronous you mean equivalent timing to custom elements, not equivalent timing to mutation events? Why is mutation observer timing insufficient?

For the same reasons node reactions are sync for CE callbacks. you want to know when the element is connected/disconnected to take immediate action, from registration, to coordinations with parent, to any other thing that all these libraries and frameworks are doing today synchronous as part of the diffing algo. Good example here is the snabbdom library that powers VUE and LWC, it is one of those popular diffing libs that provides tons of sync hooks for the operations on the dom.

> Why would you need to polyfill/emulate custom elements at this point?

Polyfilling is always handy, we don't know what the future will bring :), one example here is the scoped registry, that will take a while to take shape, but having this feature in place means that we can emulate some of the custom elements reactions without registering a real CE. For the record, we do that today in LWC, but it requires many hacks to control the connect-ness, so I might be bias. But in general, having a way to emulate these low level reactions seems like a good feature to have.

**Comment by wessberg** [comment]
@caridy, if we are going to pursue something that guarantees CE reaction timing for any given node, then I suggest opening a separate thread with such a proposal, because this thread so far has been centered around primarily `MutationObserver` which is asynchronous in it's timing. All of the Observer APIs are asynchronous, so I would argue that a sync solution would require a different approach such as callbacks or events, but in regards to events, we've already been down that road.

**Comment by caridy** [comment]
@wessberg I'm fine with a different issue. The only reason why we keep debating this here is because in the last F2F meeting we discussed this topic (from the agenda), and the point of contention was exactly what you mentioned, sync vs async.

As for the perf penalties, I will like to get more clarification from implementers. In theory, and by looking at the spec, the node reaction mechanism is in place for all elements, we only tap into them for custom elements today.  The old mutation events are not necessarily a good reference, and they semantics are confusing, but certainly something to keep in mind during this discussions. @rniwa do you want me to open another issue for the sync proposal?

**Comment by rniwa** [comment]
Let's not open a new issue. We should figure out the best way to observe the connected-ness of a node in this issue.

**Comment by wessberg** [comment]
@rniwa alright. To avoid stalling, how about we investigate whether synchronous reactions are feasible first?

@caridy, you say that you would like get feedback from implementors in regards to the perf implications of sync reactions. I think that is a great idea since it helps us understand if a sync solution is feasible. If implementors will simply block this proposal out of performance concerns and lessons learned from Mutation Events, that would be a shame. Is there anyone reading this who can chime in here or page someone who can?

I think a way forward for us is to first understand what we can all agree on in regards to goals/non-goals.

Is everone in agreement that:
- We're concerned with the connectedness of any node, and not just elements?
- We want multiple observers to be able to observe the connectedness of a node simultaneously?

**Comment by caridy** [comment]
hey implementers, any feedback on the implications of sync reactions?

@wessberg as for your two bullet points, my assumption is that yes, we want to also observe nodes, and yes, we want multiple actors to observe the connectedness of the node at once.

**Comment by Jack-Works** [comment]
If I understand this correctly, connected-ness means if a Node appears in the document or not. If so, I'd like to share my use case.
I have a `span` element and I want to mount my application on it when the `span` has been inserted in the document (I cannot control when it will be mounted). If I mount the application on the `span` when it is disconnected, it will throw an error cause most of our code doesn't assume that.

Now I have two choices, attach an event listener of deprecated `DOMNodeInserted` on the `span`, or add a `childList` `MutationObserver` on the whole document then use `document.contains`. I don't know which one hurts the performance mostly and they both don't seem like a good option.

**Comment by Jack-Works** [comment]
> Why can you not use a custom element?

Our application is a Web Extension, which `customElements` registry is no available for us.

![image](https://user-images.githubusercontent.com/5390719/119621261-7b3d6800-be38-11eb-981e-4def04e28610.png)

Web extension cannot register custom elements in content scripts.

**Comment by WebReflection** [comment]
@Jack-Works I think [this polyfill](https://github.com/ungap/custom-elements#readme) would solve your issue. Alternatively, [element-notifier](https://github.com/WebReflection/element-notifier#readme) seems to address this exact issue too.

**Comment by jfbrennan** [comment]
Just want to add my use case:

```html
<template>
  <p>My Custom Element with a menu</p>
  <button onclick="toggleMenu">Toggle</button>
  <x-menu>...</x-menu>
</template>
```

That template gets used in a Custom Element where I also call a utility function that takes the custom element as its one param, searches all its children looking for any `onevent` attributes which came from the template, and replaces them by registering an event listener - `element.addEventListener('click', e => this.toggleMenu(e))` - and removes the event property.

I'd like this utility function to also remove the listeners when the custom element is disconnected.

**Comment by LeaVerou** [comment]
Not sure if it has been suggested, but one way to make this plus a few more use cases _easier_ would be to add an option to monitor parent changes. It's simpler than the original problem, and in some ways more general.
It would still be annoying to monitor connected-ness, but a lot less so (a utility function instead of a whole library).

**Comment by FrameMuse** [comment]
The u

**Comment by FrameMuse** [comment]
## Use case

I guess the main use case would be removing subscriptions. In my case, I have a (library based) "component" that saves reference to a constructed element, the component must support callbacks for engaging scripts when the element is connected and pause/suspend those scripts when it's disconnected.

In reality, it's a bit more complex since an element may not be disconnected while actually not being in the document. So it should take into account connection/disconnection **from a document** (either parents or the element itself is connected/disconnected).

For accurate work, when https://github.com/whatwg/dom/issues/1255 is implemented, the **document** connection events should fire or they should be accounted by a developer.

## Implementation Proposal

I see it as a separate `NodeLifeObserver` class, based on `MutationObserver`, `IntersectionObserver` signature. Though this also might be worth adding dispatched events (`addEventListener`).

```ts
type NodeLifeEntryCase =
  | "connected"
  | "disconnected"

  | "parentChanged"
  | "parentConnected"
  | "parentDisconnected"

  | "documentConnected"
  | "documentDisconnected"

interface NodeLifeEntry {
  target: Node
  case: NodeLifeEntryCase
}

class NodeLifeObserver {
  constructor(callback: (entry: NodeLifeEntry[]) => void) { }

  observe(node: Node): void
  unobserve(node: Node): void
  disconnect(): void
}
```

Additionally I need the moment when is reflowed since connection/disconnection may change the size, but you need to use this recalculated value. So I expect that e.g. `connected` event would occur **before** [relow](reflow - MDN Web Docs Glossary: Definitions of Web-related terms | MDN) and when it does occur, it would trigger additional `reflowed` event.

With this I come up with the following code that extends `NodeLifeObserver` for `Element`.

```ts
type ElementLifeEntryCase =
  | NodeLifeEntryCase
  | "reflow"


interface ElementLifeEntry {
  target: Element
  case: ElementLifeEntryCase
}

class ElementLifeObserver {
  constructor(callback: (entry: ElementLifeEntry[]) => void) { }

  observe(element: Element) { }
  unobserve(element: Element) { }
  disconnect() { }
}
```

**Comment by trusktr** [comment]
<details><summary>@FrameMuse unrelated to the original discussion, but it is actually impossible to detect when an element is finally painted:</summary>

> Currently, the moment when a element is reflowed can be accurately caught with this snippet. Obviously, I would like to have a more general approach to receiving this event.
> 
> export function onElementPainted(target: Element, callback: () => void) {
>   new ResizeObserver((_, observer) => {
>     callback()
>     observer.unobserve(target)
>   }).observe(target)
> }

That's not good enough because if some other resize observer callback changes the size (directly or indirectly) of the element, it is possible that more resize callbacks will subsequently be queued and fired. The only sure thing we can do is keep the resize observer enabled the whole life of the element, instead of unobserving it right away, otherwise some reflows may be missed.

But note that reflow is not the same as paint, and it is impossible to run logic after paint. The "paint" doesn't happen until *after all resize observer callbacks*, and it is impossible to register a callback that will run after all `ResizeObserver` callbacks are finished.



</details>

**Comment by trusktr** [comment]
> type NodeLifeEntryCase =
>   | "connected"
>   | "disconnected"
> 
>   | "parentChanged" // or `reparented`.
>   | "parentConnected"
>   | "parentDisconnected"
> 
>   | "documentConnected"

What's the need for `parentConnected` (and the other parent hooks) vs the connected/disconnected? Is it that you want to run logic when an element is connected/disconnected in an inert tree that is not in the DOM? If so, I'm not sure that this is a good idea: for example the content of a `<template>`'s `.content` document is meant to be *inert* and we should avoid running logic on those trees. The idea is that when a element is connected into the live DOM (`connected`), that's when we load stuff.

Also, from the perspective of designing something simple enough for everyone to get on board with, maybe having all of that in a single proposal is too much. Could you trim down your proposal to only this,

> type NodeLifeEntryCase =
>   | "connected"
>   | "disconnected"

without the other hooks, and without `ElementLifeObserver`? We could always expand on something later, but trying to get something like that all in at once (including `ElementLifeObserver`) would be too difficult I think. They're not bad ideas (my previous comment links to much needed similar API ideas for observing style/reflow/etc), just that we may have a better time making focused incremental changes.

**Comment by trusktr** [comment]
I want to point out that having an API for element connectedness (from the outside, not limited to internal custom element methods) will help solve brain-twisting problems like this one:

- https://github.com/whatwg/dom/issues/1105

**Comment by trusktr** [comment]
Just to clarify the OP:

> similar to the way `connectedCallback` and `disconnectedCallback` work for custom elements but unlike those at the end of the microtask

Is this describing a synchronous API (not an async API like MutationObserver which runs callbacks in a microtask)?

If so, that's what I want. A synchronous API. (We already have MutationOberver for async).

**Comment by rniwa** [comment]
> Just to clarify the OP:
> 
> > similar to the way `connectedCallback` and `disconnectedCallback` work for custom elements but unlike those at the end of the microtask
> 
> Is this describing a synchronous API (not an async API like MutationObserver which runs callbacks in a microtask)?

No. The API request here is like other mutation observer callbacks and only happens at the end of a current microtask.

**Comment by wessberg** [comment]
@trusktr 
Whether to design a synchronous or asynchronous solution has been a discussion point and likely one of the main reasons why we haven't been able to move forward in any direction so far.

I for one have argued for an _asynchronous_ approach, partly because we know from the past how `MutationEvents` was shipped and later unshipped from browsers due to concerns among browser implementers about their synchronous nature.

There are several reasons why observing the connectedness of any node is not trivially solves with mutation observers, in part because a Node may move between roots in the DOM tree. Listening for changes to child lists across every existing and future Shadow Root in the document just to track a Node is not ideal, and at least my primary motivation for wanting to see this eventually being solved by observing the node itself.

**Comment by ydaniv** [comment]
We have a use-case for this feature, probably more of the same, but the ability to directly watch the connectedness of an element is the only feature we're missing. We have an interactions library that's agnostic to the rest of the code, and especially to frameworks, which may alter the DOM. 
Since not everything is yet (or ever) possible with CSS, everything the library touches must be observed. 
Currently we have to wrap every element that takes part with a custom element just for this end.

**Comment by wessberg** [comment]
@dead-claudia yes, thanks for clarifying this to prevent potential misunderstandings. I'm fully aligned with that, and I think there is strength in that it follows the same design principles as existing observers such as `MutationObserver`, `ResizeObserver`, and `IntersectionObserver`. 

E.g. something like this would be ideal to me:

```javascript
const observer = new ConnectionObserver(entries => {
	for (const {connected, target} of entries) {
		console.log("target:", target);
		console.log("connected:", connected);
	}
});

// Observe 'someElement' for connectedness (synchronous)
observer.observe(someElement);

// Eventually disconnect the observer (synchronous)
observer.disconnect();
```

I made a little proof of concept of that [here](https://github.com/wessberg/connection-observer), which works well, but requires patching `Element.prototype.attachShadow`, which is unfortunate.

**Comment by FrameMuse** [comment]
@trusktr After I published the proposal of mine I thought about it for some longer and I agree that there is less sense in listening for parent connection. I will edit my proposal, while it actually still makes sense to listen for parent connection if we don't have https://github.com/whatwg/dom/issues/736 resolved.

**Comment by FrameMuse** [comment]
@wessberg That's just awesome man! Nice poc, thank you so much

**Comment by LeaVerou** [comment]
Why would we need a separate observer for connectedness? It seems to fit right in the scope of `MutationObserver`.

**Comment by FrameMuse** [comment]
@LeaVerou Good question, I have an example demonstrating the complications, I will share it when prepared.

**Comment by justinfagnani** [comment]
It seems pretty straightforward and web-compatible to me to add a new `connection: boolean` property to `MutationObserver.observe()` options, when when true enables a new MutationRecord type of `connection`.

**Comment by dead-claudia** [comment]
> Why would we need a separate observer for connectedness? It seems to fit right in the scope of `MutationObserver`. 

Simply doing `.observe(document.documentElement, {childList: true, subtree: true})` and similar for shadow roots is sufficient for detecting connectedness changes. And that paired with a weak map would even let you put together your own unsubscribe mechanism.

This code implements liveness observers for all but children of declarative shadow roots:

```js
let observeShadow
const observers = new Set()
const registry = new FinalizationRegistry(held => {
    observers.delete(held)
})
const oldAttachShadow = Element.prototype.attachShadow
Object.defineProperty(Element.prototype, "attachShadow", {
    value() {
        const root = oldAttachShadow.apply(this, arguments)
        for (const weak of observers) {
            const observer = weak.get()
            if (observer) observeShadow(observer, root)
        }
        return root
    },
})

class LivenessObserver {
    #callback
    #observed = new WeakSet()
    #observer = new MutationObserver(records => {
        this.#check(records)
    })

    static {
        observeShadow = (observer, root) => {
            observer.#observer.observe(root, {
                childList: true,
                subtree: true,
            })
        }
    }

    constructor(callback) {
        this.#callback = callback
        observeShadow(this.#observer, document.documentElement)
        const weak = new WeakRef(this)
        observers.add(weak)
        registry.register(this, weak)
    }

    #check(records) {
        for (const r of records) {
            for (const n of r.removedNodes) {
                if (!this.#observed.has(n)) continue
                try {
                    this.#callback.removed?.(n)
                } catch (e) {
                    reportError(e)
                }
            }

            for (const n of r.addedNodes) {
                if (!this.#observed.has(n)) continue
                try {
                    this.#callback.added?.(n)
                } catch (e) {
                    reportError(e)
                }
            }
        }
    }

    observe(node) {
        this.#observed.add(node)
    }

    disconnect() {
        this.#observed.clear()
    }
}
```

Will note that this is probably not as fast as it could be.

**Comment by dead-claudia** [comment]
Back to the original feature request, I propose a minimal compromise: add a `liveness: true` flag to `MutationObserver`.

[Chromium](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/dom/node.h;l=712;drc=f4cc0c66584f7c6e05a72a3e2ad022df4ff833ba), [Firefox](https://hg-edge.mozilla.org/mozilla-central/file/c606c420560736c2182e2664024067058dd497ae/dom/base/nsINode.h#l811), and [Safari](https://github.com/WebKit/WebKit/blob/62a02c8c5d58aa668ea06379593b65c94b860179/Source/WebCore/dom/Node.h#L430) all three track this via a bit flag on the node itself. And it already needs recursively checked anyways to invoke custom element reactions, so it's not hard to just add a stack of mutation observers to possibly invoke with each node during the pass.

Unobserving nodes is better discussed in https://github.com/whatwg/dom/issues/126. Please, let's keep this on-topic, to avoid spamming everyone's notifications. (I'll admit some fault in contributing to the problem.)

**Comment by FrameMuse** [comment]
@justinfagnani Maybe I got you wrong, is this how it should work?
```js
const observer = new MutationObserver(entries => console.log(entries[0].connection))

const node = new Text("123")
observer.observe(node, { connection: true })

document.body.append(node) // => true
node.remove() // => false
```

Then I disagree. I personally think that connection is not mutation of a connected node (but a parent or document), it would be ambiguous to add it this way. The current `MutationObserver` is very on point - observe listed mutations **OF** these nodes. But in case of adding `connection` option, it goes a long way: observe mutations of all nodes, to tell if this is related to the observed node and see if it's connected or not.

So if it's already doing `childList` and `subtree` anyway, why making it ambiguous if we can introduce a new thing that just extends `MutationObserver`?

**Comment by dead-claudia** [comment]
@FrameMuse His `connected` ≈ my `liveness`.

And connection *is* a direct consequence of mutations, just not of the element itself. Consider that connection is when iframes and images start their initial load.

**Comment by dead-claudia** [comment]
> > Consider that connection is when iframes and images start their initial load.
> 
> I don't think the connection events should fire when HTML is parsed, only when connections done to existing DOM in memory.

@FrameMuse I'm talking about script-created elements here. [These are called on dynamic insertion as well as during initial HTML tree building.](https://dom.spec.whatwg.org/#ref-for-concept-node-post-connection-ext%E2%91%A5) And [here's the part for iframes I was referring to](https://html.spec.whatwg.org/multipage/iframe-embed-object.html#the-iframe-element:html-element-post-connection-steps).

I'm also expressly just asking for a way to observe that in-practice browser bit flag, so while in theory it's indirect, it in practice is very much direct observation of a node's mutable state.

**Comment by jamesdiacono** [comment]
Could Nodes simply emit non-bubbling "connect" and "disconnect" events? That would let us do everything that currently requires `connectedCallback` and `disconnectedCallback`.

**Comment by wessberg** [comment]
@jamesdiacono 

We've been down that path before with [MutationEvents](https://developer.mozilla.org/en-US/docs/Web/API/MutationEvent) which caused many issues and was later unshipped from browsers.

Instead I think we _may_ be able to achieve consensus that adding a `connected` (or similar wording) to `MutationObserver is the way forward? Based on recent discussion in this thread it seems like the approach most favored. And I'm also leaning towards agreeing that its better than my original idea of a dedicated observer.

**Comment by FrameMuse** [comment]
@dead-claudia Yeah, it depends how you think about it. So let's say we both understand each other. I'm just saying adding ~`connected` to `MutationObserver` is not immediately natural, but it'd be still very quick to adapt and get used to.

@wessberg I don't really mind for now, we can think about better place after it's shipped to browsers and countless experiments.

**Comment by FrameMuse** [comment]
@jamesdiacono AFAIK, `Event`s have different scheduling comparing to `MutationObserver` and they have many more properties than `MutationObserver`, they might be useful, but as this topic is already associated with `MutationObserver`, which historically had a problematic version of `MutationEvent` - seems olds won't accept something related to go this path again.

**Comment by LeaVerou** [comment]
Actually, I’m not sure if observing connectedness is what we need. Consider this:

```js
let parser = new DOMParser();
let doc = (new DOMParser).parseFromString("<p>Hi</p>", "text/html");
let p = doc.querySelector("p");
console.log(p.isConnected); // true
```

It seems that instead what most use cases need is to observe (a) `ownerDocument` changes and (b) `parentNode` changes. 

Then we can get all the existing custom element lifecycle hooks:
- `connectedCallback()` → `ownerDocument` goes from `null` to a non-null value
- `disconnectedCallback()` → `ownerDocument` changes from a non-null value to `null`
- `adoptedCallback()` → `ownerDocument` changes from a non-null value to a different non-null value. 
- `connectedMoveCallback()` → `parentNode` changes from a non-null value to another non-null value, while `ownerDocument` remains the same

Plus introducing such properties to `MutationObserverInit` is also much more in line with its existing API where options describe the lower-level tree variables. Connectedness is a derived variable.

**Comment by rniwa** [comment]
> It seems that instead what most use cases need is to observe (a) `ownerDocument` changes and (b) `parentNode` changes.
> 
> Then we can get all the existing custom element lifecycle hooks:
> 
> * `connectedCallback()` → `ownerDocument` goes from `null` to a non-null value
> * `disconnectedCallback()` → `ownerDocument` changes from a non-null value to `null`

`ownerDocument` is never `null` so this won't work.

**Comment by LeaVerou** [comment]
Ah, right, d'uh! 🤦🏽‍♀️
Now I’m wondering if we need a property that _does_ work like this… 😅 
Basically something like:

```js
get document () {
	let root = this.getRootNode({ composed: true });
	return root.nodeType === Node.DOCUMENT_TYPE ? root : null;
}
```

Whether such a DOM property exists or not, we _could_ introduce a `document` option that works this way.
Or, alternatively, a `root` option that lets you observe changes to either the composed root, the tree root, or both.
That actually seems to be even more broadly useful than what I was proposing above.

Observing `ownerDocument` still seems useful for implementing `adoptedCallback`, but clearly not for anything else.

**Comment by FrameMuse** [comment]
@LeaVerou I think you misinterpret it, with `DOMParser` you create a brand new document, thus you have all the elements are "connected". `isConnected` is true only when it's connected to a document, so

```js
div = document.createElement("div")
div.append(123)
div.childNodes[0].isConnected // => false

document.body.append(div)
div.childNodes[0].isConnected // => true
```

though `ownerDocument` will stay the same.

So why we need connection (`isConnected`) observer? When an element is disconnected, we want any associated subscriptions to be unsubscribed and when it's connected - subscribe back. Another use-case is to enforce placement of an element as you can track where it's connected/disconnected from and add it back - this particular use-case can solve fundamental issues for "fragments".
